File: snippets/ptr/1-alloc-int/cpp/2-ariphm.2.cpp
#include <stdio.h>
#include <iostream>

int main() {
  int arr[3] = {10, 20, 30};
  int* ptr = arr;

  // --- Addition
  int* ptr_add = ptr + 2;
  std::cout << ptr_add;

  // --- Derefernce
  int i0 = *ptr;
  std::cout << i0;

  // --- Combined: Addition + Dereference
  int i2 = ptr[2];
  // OR
  // int i2 = *(ptr + 2);
  std::cout << i2;
}

File: snippets/ptr/1-alloc-int/rust/2-ariphm2.rs


use std::hint::black_box;

fn mainn2(n: usize) {
    let arr: [i32; 3] = [10, 20, 30];
    let ptr: *const i32 = arr.as_ptr();
    black_box(ptr);

    // --- Addition
    unsafe {
        black_box(n);
        let ptr_add: *const i32 = ptr.add(n);
        black_box(ptr_add);
    }

    // --- Dereference
    unsafe {
        let mut i0: i32 = ptr.read();
        black_box(i0);
        // OR
        // let mut i0: i32 = *ptr;
    }

    // --- Combined: Addition + Dereference
    unsafe {
        let i2: i32 = *ptr.add(2);
        black_box(i2);
    }
}

fn main() {
    let s = String::from("2");
    let s2 = format!("{:}", s.parse::<usize>().unwrap());
    let n = s2.parse::<usize>().unwrap();
    mainn2(n);
}
